rclpy/rclpy/wait_set.py

```diff
@@ -0,0 +1,156 @@
+# Copyright 2017 Open Source Robotics Foundation, Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import threading
+
+from rclpy.impl.implementation_singleton import rclpy_implementation as _rclpy
+
+
+class WaitSet:
+    """Provide conveneint methods and destroy the wait set when garbage collected."""
+
+    def __init__(self):
+        # List of entity pointers (the python integer, not the PyCapsule) that are ready
+        self._ready_pointers = []
+
+        # maps pointers (integers) to handles (PyCapsule)
+        self._subscriptions = {}
+        self._guard_conditions = {}
+        self._timers = {}
+        self._clients = {}
+        self._services = {}
+
+        # Set when the wait set needs to be built or rebuilt
+        self._needs_building = True
```

Comment:
I couldn't find any line where `_needs_building` was set to `False`. Am I missing something?

Reply:
Oops, no the class is missing something

Reply:
The class only worked because it was missing something. I misunderstood how the wait set is used, but that should be fixed by 0cf2aebf61566da6fcb5b51952d5cd70b082171c

Reply:
Looks indeed better :+1: